Sources/Basic/TerminalController.swift

```diff
@@ -0,0 +1,136 @@
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import libc
+import func POSIX.getenv
+
+/// A class to have better control on tty output streams: standard output and standard error.
+/// Allows operations like cursor movement and colored text output on tty.
+public final class TerminalController {
+
+    /// Terminal color choices.
+    public enum Color {
+        case noColor
+
+        case red
+        case green
+        case yellow
+        case cyan
+
+        case white
+        case black
+        case grey
+
+        /// Returns the color code which can be prefixed on a string to display it in that color.
+        fileprivate var string: String {
+            switch self {
+                case .noColor: return ""
+                case .red: return "\u{001B}[31m"
+                case .green: return "\u{001B}[32m"
+                case .yellow: return "\u{001B}[33m"
+                case .cyan: return "\u{001B}[36m"
+                case .white: return "\u{001B}[37m"
+                case .black: return "\u{001B}[30m"
+                case .grey: return "\u{001B}[30;1m"
+            }
+        }
+    }
+
+    /// Pointer to output stream to operate on.
+    private var stream: LocalFileOutputByteStream
+
+    /// Width of the terminal.
+    public let width: Int
+
+    /// Code to clear the line on a tty.
+    private let clearLineString = "\u{001B}[2K"
+
+    /// Code to end any currently active wrapping.
+    private let resetString = "\u{001B}[0m"
+
+    /// Code to make string bold.
+    private let boldString = "\u{001B}[1m"
+
+    /// Constructs the instance if the stream is a tty.
+    public init?(stream: LocalFileOutputByteStream) {
+        // Make sure this file stream is tty.
+        guard isatty(fileno(stream.fp)) != 0 else {
+            return nil
+        }
+        width = TerminalController.terminalWidth() ?? 80 // Assume default if we are not able to determine.
+        self.stream = stream
+    }
+
+    /// Tries to get the terminal width first using COLUMNS env variable and
+    /// if that fails ioctl method testing on stdout stream.
+    ///
+    /// - Returns: Current width of terminal if it was determinable.
+    public static func terminalWidth() -> Int? {
+        // Try to get from enviornment.
+        if let columns = POSIX.getenv("COLUMNS"), let width = Int(columns) {
+            return width
+        }
+
+        // Try determining using ioctl.
+        var ws = winsize()
+        if ioctl(1, UInt(TIOCGWINSZ), &ws) == 0 {
+            return Int(ws.ws_col)
+        }
+        return nil
+    }
+
+    /// Flushes the stream.
+    public func flush() {
+        stream.flush()
+    }
+
+    /// Clears the current line and moves the cursor to beginning of the line..
+    public func clearLine() {
+        stream <<< clearLineString <<< "\r"
+        flush()
+    }
+
+    /// Moves the cursor y columns up.
+    public func moveCursor(y: Int) {
```

Comment:
BTW, we should probably find a better label here, if this is relative. It reads like an absolute position. I don't have any particular suggestion, the best I could come up with is `moveCursor(deltaY y:)` or `moveCursor(up y: Int)`.